src/sys/windows/afd.rs

```diff
@@ -0,0 +1,184 @@
+
+        unsafe {
+            let mut objname = UNICODE_STRING {
+                // Lengths are calced in bytes
+                Length: (afd_helper_name.len() * 2) as u16,
+                MaximumLength: (afd_helper_name.len() * 2) as u16,
+                Buffer: afd_helper_name.as_mut_ptr(),
+            };
+            let mut afd_helper_attributes = OBJECT_ATTRIBUTES {
+                Length: size_of::<OBJECT_ATTRIBUTES>() as ULONG,
+                RootDirectory: null_mut() as HANDLE,
+                ObjectName: &mut objname,
+                Attributes: 0 as ULONG,
+                SecurityDescriptor: null_mut() as PVOID,
+                SecurityQualityOfService: null_mut() as PVOID,
+            };
+            let status = NtCreateFile(
+                &mut afd_helper_handle,
+                SYNCHRONIZE,
+                &mut afd_helper_attributes,
+                &mut iosb,
+                null_mut(),
+                0 as ULONG,
+                FILE_SHARE_READ | FILE_SHARE_WRITE,
+                FILE_OPEN,
+                0 as ULONG,
+                null_mut(),
+                0 as ULONG,
+            );
+            if status != STATUS_SUCCESS {
+                return Err(io::Error::from_raw_os_error(
+                    RtlNtStatusToDosError(status) as i32
+                ));
+            }
+            let fd = File::from_raw_handle(afd_helper_handle as RawHandle);
+            let afd = Afd { fd };
+            let token = NEXT_TOKEN.fetch_add(1, Ordering::Relaxed) + 1;
+            cp.add_handle(token, &afd.fd)?;
+            match SetFileCompletionNotificationModes(
+                afd_helper_handle,
+                FILE_SKIP_SET_EVENT_ON_HANDLE,
+            ) {
+                0 => Err(io::Error::last_os_error()),
+                _ => Ok(afd),
+            }
+        }
+    }
+
+    pub fn poll(
+        &self,
+        info: &mut AfdPollInfo,
+        iosb: &mut IO_STATUS_BLOCK,
+        apccontext: PVOID,
+    ) -> io::Result<bool> {
+        unsafe {
+            let info_ptr: PVOID = info as *mut _ as PVOID;
+            iosb.u.Status = STATUS_PENDING;
+            let status = NtDeviceIoControlFile(
+                self.fd.as_raw_handle(),
+                null_mut(),
+                None,
+                apccontext,
+                iosb,
+                IOCTL_AFD_POLL,
+                info_ptr,
+                size_of::<AfdPollInfo>() as u32,
+                info_ptr,
+                size_of::<AfdPollInfo>() as u32,
+            );
+            match status {
+                STATUS_SUCCESS => Ok(true),
+                STATUS_PENDING => Ok(false),
+                _ => Err(io::Error::from_raw_os_error(
+                    RtlNtStatusToDosError(status) as i32
+                )),
+            }
+        }
+    }
+
+    pub fn cancel(&self, iosb: &mut IO_STATUS_BLOCK) -> io::Result<()> {
+        unsafe {
+            if iosb.u.Status != STATUS_PENDING {
```

Comment:
The IO_STATUS_BLOCK can be updated by the kernel any time when the operation is pending.
So it should probably be wrapped in an UnsafeCell and be pinned.

Reply:
UnsafeCell has no Deref trait implemented therefore cannot be pinned. I used the Box instead.

Reply:
If you can guarantee that certain memory doesn't move it can be pinned, then you don't need to use `Box`. But these kind of optimisations can be done later.

Reply:
Uploaded new code. Please check